src/scheduler.cpp

```diff
@@ -0,0 +1,209 @@
+/**
+ *  @file   scheduler.cpp
+ *  @brief  Contains definitions for functions defined in @ref scheduler.hpp
+ *  @author Charlie Kushelevsky (ckushelevsky@ucsd.edu)
+ *  @version 1
+ */
+
+#ifndef TEST_VERSION
+#include "Particle.h"
+#else
+#include "scheduler_test_system.hpp"
+#endif
+#include "cli/conio.hpp"
+#include "ensembles.hpp"
+#include "scheduler.hpp"
+#include "consts.hpp"
+
+
+ /**
+  * @brief Initializes ensembles within schedule table.
+  *
+  *
+  * @param pDeployment contains schedule table, see @ref scheduler.hpp
+  * @param startTime sets start time as given in RideTask.cpp
+  *
+  */
+void SCH_initializeSchedule(DeploymentSchedule_t* pDeployment,
+                            system_tick_t startTime)
+{
+    uint32_t lastEndTime = 0;
+    while (pDeployment->init)
+    {
+        pDeployment->deploymentStartTime = startTime;
+        pDeployment->lastMeasurementTime = 0;
+        pDeployment->measurementCount = 0;
+        lastEndTime += pDeployment->maxDuration;
+
+        pDeployment->init(pDeployment);
+        pDeployment++;
+    }
+}
+
+
+
+/**
+ * @brief Retrieves the next scheduled event
+ *
+ * This function iterates through a schedule table to find the event that
+ * should run next based on the current system time.
+ * It ensures that no overlapping events are scheduled by checking with
+ * @ref SCH_willOverlap().
+ *
+ * @see tests/gtest.cpp for intended behavior
+ *
+ * @param scheduleTable Pointer to the first element of an array of
+ * DeploymentSchedule_t, which contains the scheduling information.
+ * @param p_nextEvent Pointer to a pointer where the next event to be executed
+ * will be stored.
+ * @param p_nextTime Pointer to where the time for the next event will
+ * be stored.
+ *
+ * @return Returns SUCCESS if a suitable event is found,
+ * TASK_SEARCH_FAIL otherwise.
+ */
+
+uint32_t SCH_getNextEvent(DeploymentSchedule_t* scheduleTable,
+                DeploymentSchedule_t** p_nextEvent, system_tick_t* p_nextTime)
+{
+
+    uint32_t minNextTime = UINT32_MAX;
+    DeploymentSchedule_t* nextEvent = nullptr;
+    system_tick_t currentTime = millis();
```

Comment:
It may make sense for `SCH_getNextEvent` to instead take `currentTime` as an argument instead of pulling it from `millis`.